src/BloomExe/Extensions.cs

```diff
@@ -11,19 +11,44 @@ public static string ToLocalhost(this string fileName)
 			// don't do this if it is done already
 			if (fileName.StartsWith(ServerBase.PathEndingInSlash)) return fileName;
 
-			// BL-117, PH: With the newer xulrunner, javascript code with parenthesis in the URL is not working correctly.
-			fileName = fileName.Replace("(", "%28").Replace(")", "%29");
-
-			return ServerBase.PathEndingInSlash + fileName.Replace(":", "%3A").Replace("\\", "/");
+			return ServerBase.PathEndingInSlash + fileName.EscapeCharsForHttp().Replace(System.IO.Path.DirectorySeparatorChar, '/');
 		}
 
 		public static string FromLocalhost(this string uri)
 		{
 			if (uri.StartsWith(ServerBase.PathEndingInSlash))
-				uri = uri.Substring(ServerBase.PathEndingInSlash.Length).Replace("%3A", ":");
+				uri = uri.Substring(ServerBase.PathEndingInSlash.Length).UnescapeCharsForHttp();
 			return uri;
 		}
 
+		/// <summary>
+		/// Escapes a number of characters that need it for url/http processing.  A much larger
+		/// number could be handled, but the general case (escape all non-ASCII chars as well
+		/// as several more ASCII chars) is much more complicated, and the following appears
+		/// to suffice for our needs in communicating to our own localhost processor.
+		/// </summary>
+		public static string EscapeCharsForHttp(this string fileName)
+		{
+			fileName = fileName.Replace("%","%25");
+
+			// BL-117, PH: With the newer xulrunner, javascript code with parenthesis in the URL is not working correctly.
+			fileName = fileName.Replace("(", "%28").Replace(")", "%29");
+
+			return fileName.Replace(":", "%3A").Replace("#","%23").Replace("?","%3F");
```

Comment:
? is valid on Linux (although I would hope it wouldn't end up in file/directory names).  But it's a special character in http processing of urls as well ad #.